src/appservice/Api.ts

```diff
@@ -0,0 +1,202 @@
+import request from "request";
+import express from "express";
+import * as bodyParser from "body-parser";
+import { MjolnirManager } from "./MjolnirManager";
+import * as http from "http";
+
+/**
+ * This provides a web api that is designed to power the mjolnir widget https://github.com/matrix-org/mjolnir-widget.
+ */
+export class Api {
+    private httpdConfig: express.Express = express();
+    private httpServer?: http.Server;
+
+    constructor(
+        private homeserver: string,
+        private mjolnirManager: MjolnirManager,
+    ) {}
+
+    /**
+     * Resolves an open id access token to find a matching user that the token is valid for.
+     * @param accessToken An openID token.
+     * @returns The mxid of the user that this token belongs to or null if the token could not be authenticated.
+     */
+    private resolveAccessToken(accessToken: string): Promise<string|null> {
+        return new Promise((resolve, reject) => {
+            request({
+                url: `${this.homeserver}/_matrix/federation/v1/openid/userinfo`,
+                qs: { access_token: accessToken },
+            }, (err, homeserver_response, body) => {
+                if (err) {
+                    reject(null);
+                }
+
+                let response: { sub: string};
+                try {
+                    response = JSON.parse(body);
+                } catch (e) {
+                    reject(null);
+                    return;
+                }
+
+                resolve(response.sub);
+            });
+        });
+    }
+
+    public async close(): Promise<void> {
+        await new Promise((resolve, reject) => {
+            if (!this.httpServer) {
+                throw new TypeError("Server was never started");
+            }
+            this.httpServer.close(error => error ? reject(error) : resolve(undefined))
+        });
+    }
+
+    public start(port: number) {
+        if (this.httpServer) {
+            throw new TypeError("server already started");
+        }
+        this.httpdConfig.use(bodyParser.json());
+
+        this.httpdConfig.get("/get", this.pathGet.bind(this));
+        this.httpdConfig.get("/list", this.pathList.bind(this));
+        this.httpdConfig.post("/create", this.pathCreate.bind(this));
+        this.httpdConfig.post("/join", this.pathJoin.bind(this));
+
+        this.httpServer = this.httpdConfig.listen(port);
+    }
+
+    /**
+     * Finds the management room for a mjolnir.
+     * @param req.body.openId An OpenID token to verify the send of the request with.
```

Comment:
What are we verifying, exactly?
Do I understand correctly that only the owner of the Mjölnir can call this `GET`? What's the rationale?

Reply:
sorry yes, somehow I wrote something terrible in that docline

Reply:
and copy pasted it everywhere :(